src/Protocol/RecipeMapper.cpp

```diff
@@ -0,0 +1,104 @@
+#include "../Globals.h"
+#include "RecipeMapper.h"
+#include "../Root.h"
+
+cRecipeMapper::cRecipeMapper(void)
+{
+	cFile f;
+	if (!f.Open("Protocol/1.12.2/base.recipes.txt", cFile::fmRead))
+	{
+		LOGWARNING("Cannot open file \"Protocol/1.12.2/base.recipes.txt\", no recipe book recipes will be available!");
+		return;
+	}
+	AString Everything;
+	if (!f.ReadRestOfFile(Everything))
+	{
+		LOGWARNING("Cannot read file \"Protocol/1.12.2/base.recipes.txt\", no recipe book recipes will be available!");
+		return;
+	}
+	f.Close();
+
+	// Split it into lines, then process each line as a single recipe:
+	AStringVector Split = StringSplit(Everything, "\n");
+	m_ProtocolVersionMap["1.12.2"] = {};
+	auto RecipeNameMap = cRoot::Get()->GetCraftingRecipes()->getRecipeNameMap();
+
+	int LineNum = 1;
+	for (AStringVector::const_iterator itr = Split.begin(); itr != Split.end(); ++itr, ++LineNum)
+	{
+		// Remove anything after a '#' sign and trim away the whitespace:
+		AString Recipe = TrimString(itr->substr(0, itr->find('#')));
+		if (Recipe.empty())
+		{
+			// Empty recipe
+			continue;
+		}
+		AddRecipeLine("1.12.2", LineNum, Recipe, RecipeNameMap);
+	}
+	LOG("Loaded 1.12.2 %zu recipe book", m_ProtocolVersionMap["1.12.2"].size(), LineNum);
+}
+
+
+
+
+
+cRecipeMapper::~cRecipeMapper()
+{
+}
+
+
+
+
+
+void cRecipeMapper::AddRecipeLine(const AString & a_ProtocolVersion, int a_LineNum, const AString & a_RecipeLine, std::map<AString, UInt32> RecipeNameMap)
+{
+	AStringVector Sides = StringSplit(a_RecipeLine, " ");
+	UInt32 Id;
+	if (Sides.size() != 2)
+	{
+		LOGINFO("Recipe incompletely configured %s", a_RecipeLine);
+	}
+	StringToInteger<UInt32>(Sides[0].c_str(), Id);
+	const UInt32 RecipeIndex = RecipeNameMap[Sides[1]];
+	m_ProtocolVersionMap[a_ProtocolVersion].emplace(Id, RecipeIndex);
+}
+
+
+
+
+
+UInt32 cRecipeMapper::GetProtocolRecipeId(UInt32 a_RecipeId, UInt32 a_ProtocolVersion)
+{
+	auto ProtocolMap = m_ProtocolVersionMap.find(cRoot::Get()->GetProtocolVersionTextFromInt(static_cast<int>(a_ProtocolVersion)));
+	if (ProtocolMap == m_ProtocolVersionMap.end())
+	{
+		return 0;
```

Comment:
Hm, I guess I need help with this. I included the change to `std::optional<UInt32>` (https://github.com/cuberite/cuberite/pull/4493/commits/e164415a18b349fed9ff4a2615587db65eba7cad#diff-c8b569fcdb964b6253967e43533f707eR11)

but running `make` throws: `cuberite/src/Items/../Protocol/RecipeMapper.h:11:7: error: ‘optional’ in namespace ‘std’ does not name a template type
  std::optional<UInt32> GetProtocolRecipeId(UInt32 a_RecipeId, UInt32 a_ProtocolVersion);`

The branch is rebased on master. I did a fresh clone (to avoid any old compile stuff), `cmake .`, `make`.

Reading it up sounds like C++17 is not enabled,
Cmake says the internet is lying: `-- Using c++ standard c++17`

What am I doing wrong? Sorry, I'm a bit rusty with C++

Reply:
Nevermind, ignore me. Travis is happy, so am I.

Seems like I have to fix my local build environment ... somehow.

Reply:
Ok, not that easy: It seems like appveyor is not yet configured properly: `C:\projects\cuberite\src\Protocol/RecipeMapper.h(12,7): error C2039: 'optional': is not a member of 'std' ` not sure if I can do anything here.